app/src/lib/stores/helpers/branch-pruner.ts

```diff
@@ -0,0 +1,143 @@
+import { Repository } from '../../../models/repository'
+import { RepositoriesStore } from '../repositories-store'
+import { RepositoryStateCache } from '../repository-state-cache'
+import { Branch, BranchType } from '../../../models/branch'
+import { GitStoreCache } from '../git-store-cache'
+import { getMergedBranches } from '../../git'
+import { fatalError } from '../../fatal-error'
+import { enableBranchPruning } from '../../feature-flag'
+
+/** Check if a repo needs to be pruned at least every 4 hours */
+const BackgroundPruneMinimumInterval = 1000 * 60 * 60 * 4
+
+export class BranchPruner {
+  private timer: NodeJS.Timer | null = null
+
+  public constructor(
+    private readonly repository: Repository,
+    private readonly gitStoreCache: GitStoreCache,
+    private readonly repositoriesStore: RepositoriesStore,
+    private readonly repositoriesStateCache: RepositoryStateCache,
+    private readonly onPruneCompleted: (repository: Repository) => Promise<void>
+  ) {}
+
+  public async start() {
+    if (enableBranchPruning() === false) {
+      return
+    }
+
+    if (this.timer !== null) {
+      fatalError(
+        `A background prune task is already active and cannot begin pruning on ${
+          this.repository.name
+        }`
+      )
+    }
+
+    await this.pruneLocalBranches()
+    this.timer = setInterval(
+      () => this.pruneLocalBranches(),
+      BackgroundPruneMinimumInterval
+    )
+  }
+
+  public stop() {
+    if (this.timer === null) {
+      return
+    }
+
+    clearInterval(this.timer)
+    this.timer = null
+  }
+
+  private async findBranchesMergedIntoDefaultBranch(
+    repository: Repository,
+    defaultBranch: Branch
+  ): Promise<ReadonlyArray<string> | null> {
+    const gitStore = this.gitStoreCache.get(repository)
+    return (
+      (await gitStore.performFailableOperation(() =>
+        getMergedBranches(repository, defaultBranch.name)
+      )) || null
+    )
+  }
+
+  private async pruneLocalBranches(): Promise<void> {
+    if (this.repository.gitHubRepository === null) {
+      return
+    }
+
+    // Get the last time this repo was pruned
+    const lastPruneDate = await this.repositoriesStore.getLastPruneDate(
+      this.repository
+    )
+
+    // Only prune if it's been at least 24 hours since the last time
+    if (
+      !__DEV__ &&
```

Comment:
I don't think this is entirely covered by the feature flag check, as given how aggressive this might be I think you want to have situations where this doesn't run even after it's been stabilised is considered ready for production use.

Reply:
The `__Dev__` check is for me to easily test, it will be removed before merging.